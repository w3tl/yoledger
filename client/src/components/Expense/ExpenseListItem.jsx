import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { expensePropType } from './propTypes';

function ExpenseListItem({ url, expense }) {
  const { id, name } = expense;
  return (
    <li key={name}>
      {url ? (
        <Link to={{ pathname: `${url}/view`, state: { id } }}>{name}</Link>
      ) : (
        name
      )}
    </li>
  );
}

ExpenseListItem.propTypes = {
  url: PropTypes.string,
  expense: expensePropType.isRequired,
};

ExpenseListItem.defaultProps = {
  url: null,
};

export default ExpenseListItem;
