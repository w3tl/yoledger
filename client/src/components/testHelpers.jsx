import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
