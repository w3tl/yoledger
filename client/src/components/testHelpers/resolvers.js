import dateResolver from '../../../../graphql/date';
import assets from '../Asset/mockData';
import expenses from '../Expense/mockData';
import incomes from '../Income/mockData';
import { accounts, periods, budgets } from '../Budget/mockData';

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      if (id === '666') return { name: 'new account' };
      return assets.concat(expenses).concat(incomes).find(a => a.id === id);
    },
    accounts(root, { type }) {
      if (type === 'ASSET') return assets;
      if (type === 'EXPENSE') return expenses;
      if (type === 'INCOME') return incomes;
      return [];
    },
    budgets() {
      return { accounts, periods: periods.map(dateStr => new Date(dateStr)) };
    },
    budget() {
      return budgets[0];
    },
  },
  Mutation: {
    upsertBudget(root, { input: { date } }) {
      return {
        success: true,
        allocation: {
          id: date.toISOString().concat('new account'),
          account: { id: '99', name: 'new account' },
        },
      };
    },
    addAccount(root, { input }) {
      return {
        account: { id: '666', name: input.name },
      };
    },
  },
};
