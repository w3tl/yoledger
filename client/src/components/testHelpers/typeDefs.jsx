export default `
scalar Date
type Query {
  account(id: ID!): Account
  accounts(type: AccountType!): [Account]

  budgets(dateStart: Date!, count: Int = 5): Budget
  budget(date: Date!): [Allocation]
}
type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

type Budget {
  periods: [Date]
  accounts: [Account]
}

type Allocation {
  id: String!
  account: Account!
  amount: Float
  balance: Float
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  allocation: Allocation
}
`;
