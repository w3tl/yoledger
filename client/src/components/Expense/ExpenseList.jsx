import React from 'react';
import PropTypes from 'prop-types';
import ExpenseListItem from './ExpenseListItem';
import { expensePropType } from './propTypes';

function ExpenseList({ expenses, match }) {
  return (
    <ul>
      {expenses.map(account => (
        <ExpenseListItem key={account.name} url={match.url} expense={account} />))}
    </ul>
  );
}

ExpenseList.propTypes = {
  expenses: PropTypes.arrayOf(expensePropType),
  match: PropTypes.shape({
    url: PropTypes.string,
  }),
};

ExpenseList.defaultProps = {
  match: {
    url: '',
  },
  expenses: [],
};

export default ExpenseList;
