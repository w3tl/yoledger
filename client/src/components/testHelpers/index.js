import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
