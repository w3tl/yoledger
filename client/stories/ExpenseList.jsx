import React from 'react';
import { storiesOf } from '@storybook/react';

import ExpenseList from '../src/components/Expense/ExpenseList';
import expenses from '../src/components/Expense/mockData';

storiesOf('Expense/List', module)
  .add('with data', () => <ExpenseList expenses={expenses} />)
  .add('empty list', () => <ExpenseList />);
