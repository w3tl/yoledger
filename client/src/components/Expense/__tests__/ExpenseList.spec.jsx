import React from 'react';
import { shallow } from 'enzyme';
import ExpenseList from '../ExpenseList';
import ExpenseListItem from '../ExpenseListItem';
import { testRenderWithoutError } from '../../testHelpers';
import expenses from '../mockData';

jest.mock('react-router-dom', () => ({ // eslint-disable-next-line react/prop-types
  Link: ({ to, children }) => <a href={to}>{children}</a>,
}));

describe('ExpenseList', () => {
  testRenderWithoutError(<ExpenseList expenses={expenses} />);

  it('should contain ExpenseListItem', () => {
    const output = shallow(
      <ExpenseList expenses={expenses} />,
    );
    expect(output.find(ExpenseListItem)).toHaveLength(expenses.length);
  });
});
