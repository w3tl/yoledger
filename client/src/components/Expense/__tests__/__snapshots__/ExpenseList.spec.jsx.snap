// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList renders without error 1`] = `
<ul>
  <ExpenseListItem
    expense={
      Object {
        "id": "21",
        "name": "Food",
      }
    }
    key="Food"
    url=""
  />
  <ExpenseListItem
    expense={
      Object {
        "id": "22",
        "name": "Beer",
      }
    }
    key="Beer"
    url=""
  />
  <ExpenseListItem
    expense={
      Object {
        "id": "23",
        "name": "Train",
      }
    }
    key="Train"
    url=""
  />
</ul>
`;
